### PR TITLE
[JENKINS-60866] Advanced button, build health, confirmation link, URI encoding monitor

### PR DESCRIPTION
**Pull request by daniel-beck**
See [JENKINS-60866](https://issues.jenkins-ci.org/browse/JENKINS-60866).

### Testing Notes

* `confirmationLink` is e.g. the "Reload from Disk" item on `/manage`, the script controls the confirmation dialog.
* `buildHealth` is the weather column and the weather icon in the build history title, the script controls hover behavior.
* `URICheckEncodingMonitor` is only available on `/manage`. It shows fairly easily for me (who has Latin 1?) but the error I could only trigger by changing `params` in JS so that actual/expected are different. The monitor has nonstandard appearance since we styled them differently from 1.x, out of scope for this.
* `advanced` is the script that controls whether the "something was edited" message appears to the left of the button; this is trivially the case with Mailer plugin's "Send a test email" misuse of this control.

### Proposed changelog entries

* Entry 1: Issue, Human-readable Text
* ...

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: test/src/test/java/lib/layout/ConfirmationLinkTest.java
Comment (daniel-beck) [inline test/src/test/java/lib/layout/ConfirmationLinkTest.java:92]: This looks like a case we don't really need to support, @Wadeck to confirm as original author.

**Review by timja — COMMENTED** [review]
How is this related to https://issues.jenkins.io/browse/JENKINS-60881 ?

**Comment by daniel-beck** [comment]
@timja Thanks, updated.

File: core/src/main/resources/lib/hudson/buildHealth.jelly
Comment (Wadeck) [inline core/src/main/resources/lib/hudson/buildHealth.jelly:89]: :bug: This does not work due to j:parse, please see global comment.

File: test/src/test/java/lib/layout/ConfirmationLinkTest.java
Comment (Wadeck) [reply test/src/test/java/lib/layout/ConfirmationLinkTest.java:92]: ✔️ No strict need, it's expected to be a boolean, I was just more lenient with the test.

File: core/src/main/resources/lib/layout/confirmationLink/adjunct.js
Comment (Wadeck) [inline core/src/main/resources/lib/layout/confirmationLink/adjunct.js:2]: Same comment as for validateButton

**Review by Wadeck — CHANGES_REQUESTED** [review]
The missing wrapping behavior could be potentially problematic but I have no examples to demonstrate that :-)

----

🐛 Only one where I have serious doubts about is the healthReport in the widget. The adjunct seems not loaded/effective.

<details>

<summary>Working proofs</summary>
In case something was wrong but I was not aware :p

### URICheckEncodingMonitor

Screenshot_2021-05-25_080338_001

### Advanced button

Screenshot_2021-05-25_081400_001

### Validate button

Screenshot_2021-05-25_081923_001

### BuildHealth in the job table

The effect is working fine and the adjunct is inserted.

#### Hover working thanks to the added/removed class

Screenshot_2021-05-25_084151_001


#### Adjunct added as expected

Screenshot_2021-05-25_084250_001


### Confirmation link

Screenshot_2021-05-25_103020_001


</details>


<details>

<summary>Not working proofs</summary>

### 🐛 BuildHealth in the left widget

[Disclaimer] To see anything, you need to have at least one build in the history of the job you are looking, otherwise the tooltip will not exist.

The effect is working because of `:hover`, not because of the `.hover` class added/removed by the script.
The script itself does not seem to be loaded, even if using the debugging I saw the code from AdjunctsInPage#generate being called.

💡 AFAICT the problem is coming from the widget/ajax nature of that feature.

Be careful, the "desired effect" is working, but not in the way you would expect, and thus could lead to other errors in the future that will be very complicated to solve as the code will "seem to work".

#### No hover, but tooltip displayed thanks to :hover

Screenshot_2021-05-25_083523_001

#### No adjunct for BuildHealth

Screenshot_2021-05-25_083849_001

#### On ci.jenkins.io

As you can see, the hover class is added and you can even see the mouseenter/out event inlined at that point.

Screenshot_2021-05-25_084447_001

#### Debug information
I compared the confirmationLink adjunct with the buildHealth. The first one is included correctly in the page, compared to the second that is not. The difference I have seen is about the `out`, being an `HTMLWriterOutput` for confirmationLink _(and all other adjuncts)_ and an `XMLOutput` for buildHealth.

Screenshot_2021-05-25_091103_001

Screenshot_2021-05-25_091114_001

#### Debug with intelligence (headache included for free 🎁)

We are changing HTMLWriterOutput to XMLOutput in the frames at ParseTag#parseBody:

Screenshot_2021-05-25_091530_001

More precisely at [HistoryWidget/index.jelly#L28](https://github.com/jenkinsci/jenkins/blob/62546443947fafecc94a5e3b47bb9129bed8e148/core/src/main/resources/hudson/widgets/HistoryWidget/index.jelly#L28), that is then executed in [HistoryWidget/index.jelly#L62](https://github.com/jenkinsci/jenkins/blob/62546443947fafecc94a5e3b47bb9129bed8e148/core/src/main/resources/hudson/widgets/HistoryWidget/index.jelly#L62).

The problem is that AdjunctsInPage#generate is writing the script "directly" to the output, meaning in text instead of as a tag. When the ParseTag#parseBody does the endDocument, it erases the textBuffer and so, delete the just added script. To correct this behavior we should write the script as a tag, like:

Screenshot_2021-05-25_100924_001

Which, is working fine.

💡 The quick workaround is to call manually the adjunct from buildHealth *before* the f:parse (manualy tested).

</details>

**Comment by oleg-nenashev** [comment]
@Wadeck did you have a chance to re-review it?

**Review by Wadeck — APPROVED** [review]
Working fine and correctly now, thanks for the correct 👍

**Review by oleg-nenashev — APPROVED** [review]
Approved once the merge conflict is resolved

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Review by Wadeck — APPROVED** [review]
> (from Daniel) do you know whether this is fine? I do not remember 😞

> (from my previous tests) The effect is working because of :hover, not because of the .hover class added/removed by the script.

So, it's a cleanup => 👍 

Also, I checked the related CSS to this "feature", it's expecting any of :hover or `.hover`. So another cleanup task could be to remove the `.hover` part as a follow-up :) (code: [style.less#L787-L792](https://github.com/jenkinsci/jenkins/blob/21165955b19518379ed11ff7e34e36749d36fb8b/war/src/main/less/base/style.less#L787-L792)

(additional manual testing with the new BuildHistory from Jan 🎉)

**Pull request merged**
